README
This is a node js server side QR code / 2d barcode generator.
it is an extension of "QRCode for JavaScript" which Kazuhiko Arase thankfully mit licensed

the qrcode-draw.js can be used  directly as a client side lib if its appended too or included with lib/qrcode.js
to use this on the server side please require('qrcode'); =)


install:

	npm install qrcode

dependencies:

	npm install canvas

api:
	QRCode.draw(text,cb(error,canvas));
		returns node canvas object see https://github.com/LearnBoost/node-canvas for all of the cool node things you can do
		look up the canvas api for the other cool things

	QRCode.toDataURL(text,cb(error,dataURL));
		returns mime image/png data url for the 2d barcode 

	QRCode.save(path,text,cb(error,written));
		saves png to the path specified returns bytes
	
examples:
	---code---
	// simple test
	var sys = require('sys');
	var QRCode = require('qrcode');

	QRCode.toDataURL('i am a pony!',function(err,url){
		sys.print(url);
	});

	--- execute ---
	node ./tests/url.js


	for server use
		see tests/server.js

	for client side use
		open tests/clientside.html in your browser



The word "QR Code" is registered trademark of 
DENSO WAVE INCORPORATED

CHANGELOG.md
## [1.3.3](https://github.com/soldair/node-qrcode/compare/v1.3.2...v1.3.3) (2019-01-16)

fixing security vulnerabillities reported by users of snyk and `npm audit` major
versions of deps were bumped and should only impact development of qrcode.


## [1.3.2](https://github.com/soldair/node-qrcode/compare/v1.3.0...v1.3.2) (2019-01-16)


### Bug Fixes

* forgot can-promise replacement ([2514445](https://github.com/soldair/node-qrcode/commit/2514445))
* replacing can-promise ([8197d78](https://github.com/soldair/node-qrcode/commit/8197d78))
* security vulnerabillities and a ref to new Buffer ([a40c757](https://github.com/soldair/node-qrcode/commit/a40c757))



# [1.3.0](https://github.com/soldair/node-qrcode/compare/v1.2.1...v1.3.0) (2018-10-01)



## [1.2.1](https://github.com/soldair/node-qrcode/compare/v1.2.0...v1.2.1) (2018-06-06)



# [1.2.0](https://github.com/soldair/node-qrcode/compare/v1.1.0...v1.2.0) (2017-12-28)



# [1.1.0](https://github.com/soldair/node-qrcode/compare/v1.0.1...v1.1.0) (2017-12-28)



## [1.0.1](https://github.com/soldair/node-qrcode/compare/v1.0.0...v1.0.1) (2017-12-27)



# [1.0.0](https://github.com/soldair/node-qrcode/compare/v0.9.0...v1.0.0) (2017-11-08)

no breaking changes. promoting to a stable semver

# [0.9.0](https://github.com/soldair/node-qrcode/compare/v0.8.2...v0.9.0) (2017-07-22)



## [0.8.2](https://github.com/soldair/node-qrcode/compare/v0.8.0...v0.8.2) (2017-05-23)



# [0.8.0](https://github.com/soldair/node-qrcode/compare/v0.7.1...v0.8.0) (2017-03-30)



## [0.7.1](https://github.com/soldair/node-qrcode/compare/v0.7.0...v0.7.1) (2017-03-17)



# [0.7.0](https://github.com/soldair/node-qrcode/compare/v0.6.0...v0.7.0) (2017-02-27)



# [0.6.0](https://github.com/soldair/node-qrcode/compare/v0.5.0...v0.6.0) (2017-02-24)



# [0.5.0](https://github.com/soldair/node-qrcode/compare/v0.4.4...v0.5.0) (2016-09-19)



## [0.4.4](https://github.com/soldair/node-qrcode/compare/v0.4.3...v0.4.4) (2016-08-25)



## [0.4.3](https://github.com/soldair/node-qrcode/compare/v0.4.0...v0.4.3) (2016-08-16)



# [0.4.0](https://github.com/soldair/node-qrcode/compare/0.2.10...v0.4.0) (2015-09-17)



## [0.1.1](https://github.com/soldair/node-qrcode/compare/v0.1.0...v0.1.1) (2011-04-17)



# [0.1.0](https://github.com/soldair/node-qrcode/compare/0.0.3...v0.1.0) (2011-04-17)



## [0.0.3](https://github.com/soldair/node-qrcode/compare/0.0.2...0.0.3) (2011-02-27)



## 0.0.2 (2010-12-27)





lib/utils.js
var MaskPattern = require('./mask-pattern')
var Mode = require('./mode')
var GF = require('./galois-field')
var Polynomial = require('./polynomial')

module.exports = {
  PATTERN_POSITION_TABLE : [
    [],
    [6, 18],
    [6, 22],
    [6, 26],
    [6, 30],
    [6, 34],
    [6, 22, 38],
    [6, 24, 42],
    [6, 26, 46],
    [6, 28, 50],
    [6, 30, 54],    
    [6, 32, 58],
    [6, 34, 62],
    [6, 26, 46, 66],
    [6, 26, 48, 70],
    [6, 26, 50, 74],
    [6, 30, 54, 78],
    [6, 30, 56, 82],
    [6, 30, 58, 86],
    [6, 34, 62, 90],
    [6, 28, 50, 72, 94],
    [6, 26, 50, 74, 98],
    [6, 30, 54, 78, 102],
    [6, 28, 54, 80, 106],
    [6, 32, 58, 84, 110],
    [6, 30, 58, 86, 114],
    [6, 34, 62, 90, 118],
    [6, 26, 50, 74, 98, 122],
    [6, 30, 54, 78, 102, 126],
    [6, 26, 52, 78, 104, 130],
    [6, 30, 56, 82, 108, 134],
    [6, 34, 60, 86, 112, 138],
    [6, 30, 58, 86, 114, 142],
    [6, 34, 62, 90, 118, 146],
    [6, 30, 54, 78, 102, 126, 150],
    [6, 24, 50, 76, 102, 128, 154],
    [6, 28, 54, 80, 106, 132, 158],
    [6, 32, 58, 84, 110, 136, 162],
    [6, 26, 54, 82, 110, 138, 166],
    [6, 30, 58, 86, 114, 142, 170]
  ],

  G15 : (1 << 10) | (1 << 8) | (1 << 5) | (1 << 4) | (1 << 2) | (1 << 1) | (1 << 0),
  G18 : (1 << 12) | (1 << 11) | (1 << 10) | (1 << 9) | (1 << 8) | (1 << 5) | (1 << 2) | (1 << 0),
  G15_MASK : (1 << 14) | (1 << 12) | (1 << 10)  | (1 << 4) | (1 << 1),

  getBCHTypeInfo : function(data) {
    var d = data << 10;
    while (this.getBCHDigit(d) - this.getBCHDigit(this.G15) >= 0) {
      d ^= (this.G15 << (this.getBCHDigit(d) - this.getBCHDigit(this.G15) ) );  
    }
    return ( (data << 10) | d) ^ this.G15_MASK;
  },

  getBCHTypeNumber : function(data) {
    var d = data << 12;
    while (this.getBCHDigit(d) - this.getBCHDigit(this.G18) >= 0) {
      d ^= (this.G18 << (this.getBCHDigit(d) - this.getBCHDigit(this.G18) ) );  
    }
    return (data << 12) | d;
  },

  getBCHDigit : function(data) {

    var digit = 0;

    while (data != 0) {
      digit++;
      data >>>= 1;
    }

    return digit;
  },

  getPatternPosition : function(typeNumber) {
    return this.PATTERN_POSITION_TABLE[typeNumber - 1];
  },

  getMask : function(maskPattern, i, j) {
    
    switch (maskPattern) {
      
    case MaskPattern.PATTERN000 : return (i + j) % 2 == 0;
    case MaskPattern.PATTERN001 : return i % 2 == 0;
    case MaskPattern.PATTERN010 : return j % 3 == 0;
    case MaskPattern.PATTERN011 : return (i + j) % 3 == 0;
    case MaskPattern.PATTERN100 : return (Math.floor(i / 2) + Math.floor(j / 3) ) % 2 == 0;
    case MaskPattern.PATTERN101 : return (i * j) % 2 + (i * j) % 3 == 0;
    case MaskPattern.PATTERN110 : return ( (i * j) % 2 + (i * j) % 3) % 2 == 0;
    case MaskPattern.PATTERN111 : return ( (i * j) % 3 + (i + j) % 2) % 2 == 0;

    default :
      throw new Error("bad maskPattern:" + maskPattern);
    }
  },

  getErrorCorrectPolynomial : function(errorCorrectLength) {

    var a = new Polynomial([1], 0);

    for (var i = 0; i < errorCorrectLength; i++) {
      a = a.multiply(new Polynomial([1, GF.exp(i)], 0) );
    }

    return a;
  },

  getLengthInBits : function(mode, type) {

    if (1 <= type && type < 10) {

      // 1 - 9

      switch(mode) {
      case Mode.MODE_NUMBER   : return 10;
      case Mode.MODE_ALPHA_NUM  : return 9;
      case Mode.MODE_8BIT_BYTE  : return 8;
      case Mode.MODE_KANJI    : return 8;
      default :
        throw new Error("mode:" + mode);
      }

    } else if (type < 27) {

      // 10 - 26

      switch(mode) {
      case Mode.MODE_NUMBER   : return 12;
      case Mode.MODE_ALPHA_NUM  : return 11;
      case Mode.MODE_8BIT_BYTE  : return 16;
      case Mode.MODE_KANJI    : return 10;
      default :
        throw new Error("mode:" + mode);
      }

    } else if (type < 41) {

      // 27 - 40

      switch(mode) {
      case Mode.MODE_NUMBER   : return 14;
      case Mode.MODE_ALPHA_NUM  : return 13;
      case Mode.MODE_8BIT_BYTE  : return 16;
      case Mode.MODE_KANJI    : return 12;
      default :
        throw new Error("mode:" + mode);
      }

    } else {
      throw new Error("type:" + type);
    }
  },

  getLostPoint : function(qrCode) {
    
    var moduleCount = qrCode.getModuleCount();
    
    var lostPoint = 0;
    
    // LEVEL1
    
    for (var row = 0; row < moduleCount; row++) {

      for (var col = 0; col < moduleCount; col++) {

        var sameCount = 0;
        var dark = qrCode.isDark(row, col);

      for (var r = -1; r <= 1; r++) {

          if (row + r < 0 || moduleCount <= row + r) {
            continue;
          }

          for (var c = -1; c <= 1; c++) {

            if (col + c < 0 || moduleCount <= col + c) {
              continue;
            }

            if (r == 0 && c == 0) {
              continue;
            }

            if (dark == qrCode.isDark(row + r, col + c) ) {
              sameCount++;
            }
          }
        }

        if (sameCount > 5) {
          lostPoint += (3 + sameCount - 5);
        }
      }
    }

    // LEVEL2

    for (var row = 0; row < moduleCount - 1; row++) {
      for (var col = 0; col < moduleCount - 1; col++) {
        var count = 0;
        if (qrCode.isDark(row,     col    ) ) count++;
        if (qrCode.isDark(row + 1, col    ) ) count++;
        if (qrCode.isDark(row,     col + 1) ) count++;
        if (qrCode.isDark(row + 1, col + 1) ) count++;
        if (count == 0 || count == 4) {
          lostPoint += 3;
        }
      }
    }

    // LEVEL3

    for (var row = 0; row < moduleCount; row++) {
      for (var col = 0; col < moduleCount - 6; col++) {
        if (qrCode.isDark(row, col)
            && !qrCode.isDark(row, col + 1)
            &&  qrCode.isDark(row, col + 2)
            &&  qrCode.isDark(row, col + 3)
            &&  qrCode.isDark(row, col + 4)
            && !qrCode.isDark(row, col + 5)
            &&  qrCode.isDark(row, col + 6) ) {
          lostPoint += 40;
        }
      }
    }

    for (var col = 0; col < moduleCount; col++) {
      for (var row = 0; row < moduleCount - 6; row++) {
        if (qrCode.isDark(row, col)
            && !qrCode.isDark(row + 1, col)
            &&  qrCode.isDark(row + 2, col)
            &&  qrCode.isDark(row + 3, col)
            &&  qrCode.isDark(row + 4, col)
            && !qrCode.isDark(row + 5, col)
            &&  qrCode.isDark(row + 6, col) ) {
          lostPoint += 40;
        }
      }
    }

    // LEVEL4
    
    var darkCount = 0;

    for (var col = 0; col < moduleCount; col++) {
      for (var row = 0; row < moduleCount; row++) {
        if (qrCode.isDark(row, col) ) {
          darkCount++;
        }
      }
    }
    
    var ratio = Math.abs(100 * darkCount / moduleCount / moduleCount - 50) / 5;
    lostPoint += ratio * 10;

    return lostPoint;   
  }
};
